framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,311 @@
+ * #postInit()}.
+ */
+public abstract class AccumulationAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+
+    /**
+     * The canonical top annotation for this accumulation checker: an instance of the accumulator
+     * annotation with no arguments.
+     */
+    public final AnnotationMirror top;
+
+    /** The canonical bottom annotation for this accumulation checker. */
+    public final AnnotationMirror bottom;
+
+    /**
+     * The annotation that accumulates things in this accumulation checker. Must be an annotation
+     * with exactly one field named "value" whose type is a String array.
+     */
+    private final Class<? extends Annotation> accumulator;
+
+    /**
+     * Create an annotated type factory for an accumulation checker.
+     *
+     * @param checker the checker
+     * @param accumulator the accumulator type in the hierarchy. Must be an annotation with a single
+     *     argument named "value" whose type is a String array.
+     * @param bottom the bottom type in the hierarchy, which must be a subtype of {@code
+     *     accumulator}. The bottom type should be an annotation with no arguments.
+     */
+    protected AccumulationAnnotatedTypeFactory(
+            BaseTypeChecker checker,
+            Class<? extends Annotation> accumulator,
+            Class<? extends Annotation> bottom) {
+        super(checker);
+
+        this.accumulator = accumulator;
+
+        // Check that the requirements of the accumulator are met.
+        Method[] accDeclaredMethods = accumulator.getDeclaredMethods();
+        if (accDeclaredMethods.length != 1) {
+            rejectMalformedAccumulator("not have more than one argument");
+        }
+        Method value = accDeclaredMethods[0];
+        if (value.getName() != "value") {
+            rejectMalformedAccumulator("name its argument \"value\"");
+        }
+        if (!value.getReturnType().isInstance(new String[0])) {
+            rejectMalformedAccumulator("have an argument of type String[]");
+        }
+        if (((String[]) value.getDefaultValue()).length != 0) {
+            rejectMalformedAccumulator("must have the empty string array {} as its default value");
+        }
+
+        this.bottom = AnnotationBuilder.fromClass(elements, bottom);
+        this.top = createAccumulatorAnnotation(Collections.emptyList());
+
+        // Every subclass must call postInit!  This does not do so for subclasses.
+        if (this.getClass() == AccumulationAnnotatedTypeFactory.class) {
+            this.postInit();
+        }
+    }
+
+    /**
+     * Common error message for malformed accumulator annotation.
+     *
+     * @param missing what is missing from the accumulator, suitable for use in this string to
+     *     replace $MISSING$: "The accumulator annotation Foo must $MISSING$."
+     */
+    private void rejectMalformedAccumulator(String missing) {
+        throw new UserError(
```

Comment:
A UserError is when a user passed wrong arguments to the Checker Framework.  It can be corrected by changing the run-time invocation.  A bug in the source code of a checker should be thrown as a BugInCF., even if that source code was written by someone other than the main Checker Framework developers.